lib/acts_as_taggable_on_mongoid/taggable/tagged_with_query/base.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module ActsAsTaggableOnMongoid
+  module Taggable
+    class TaggedWithQuery
+      class Base
+        attr_reader :taggable_model,
+                    :tag_definition,
+                    :tag_list,
+                    :options
+
+        def initialize(tag_definition, tag_list, options)
+          @tag_definition = tag_definition
+          @tag_list       = tag_list
+          @options        = options
+        end
+
+        private
+
+        # Any is relatively simple, but All and exclude are a bit more complicated.  To make the code simpler
+        # I'm treating all of them the same.
+        # We build an aggregation of all of the matching taggables whose key_name is in the list of tags with the
+        # count of the matching key names.  We then filter on that count.
+        #
+        # * All - count == tag_list count
+        # * Any - count > 0
+        # * Exclude - count > 0 (but anything that isn't in that count)
+        def build_ids_from(count_selector)
+          where_query = tagging_query.where(:tag_name.in => tag_list)
+          build_ids_from_query(where_query, count_selector)
+        end
+
+        def build_tagless_ids_from(count_selector)
+          build_ids_from_query(tagging_query, count_selector)
+        end
+
+        def build_ids_from_query(where_query, count_selector)
+          pipeline = where_query.
+              group(_id: { taggable_id: "$taggable_id", tag_name: "$tag_name" }).
+              group(_id: "$_id.taggable_id", :count.sum => 1).
+              pipeline.
+              concat(count_selector.to_pipeline)
```

Comment:
The pipeline is to first get a list of distinct tag values per taggable_id.
We then group by taggable_ids with a count that matches the WHERE clause.

This allows us to get a count of distinct tag names for the taggable record.

The final concat adds a final where (read: HAVING) clause that causes us to get the final taggings we want.